Carry objects

NickAragua opened the pull request:
Implements cargo carrier rules from TW:261 (mechs and protomechs only) and (partly) TOAR:90 (just the part where you can pick up a lighter object with one hand only).

Includes:
- picking up cargo (mechs and protomechs only)
- dropping/unloading cargo (drop if you keep moving, unload if it's the last step)
- can't fire if loaded/unloaded cargo this turn
- can't fire torso/arm weapons if carrying cargo (can use other arm when carrying really light cargo)
- cargo will get damaged if the carrying unit is shot at
- cargo may get destroyed if carrying unit drops it on the move (even more likely when flying/jumping)
- graphical indicators and reporting of cargo-related activity
- cargo can be 0-weight and invulnerable
- if you fall, you drop the cargo

Does not include
- TSM bonus
- Loading cargo for non-mech/protomech units

A story told in a few screenshots:

![image](https://github.com/user-attachments/assets/32b08cca-f65d-464e-9446-2f585c2c74dd)

![image](https://github.com/user-attachments/assets/47bad613-c5bb-4572-acf6-561a05ce0b99)

![image](https://github.com/user-attachments/assets/87b6b788-a4a0-452a-9b6f-d616d57bb029)

![image](https://github.com/user-attachments/assets/20ee4ca5-0316-4537-b450-25c01804f3ac)

![image](https://github.com/user-attachments/assets/ee1e3756-c224-4893-9446-501f4b2bc8ac)

![image](https://github.com/user-attachments/assets/4e225a62-de56-4521-9ac5-c959296aa51b)

HammerGS commented:
Would there be any opportunity in this to handle the Lift Hoists being able to load and carry a unit and drop on the map?

NickAragua commented:
Definitely not units. I'll look into Lift Hoists/infantry being able to load cargo in a subsequent pull request.

NickAragua commented:
Not quite complete, but can be put through some initial testing/feedback.

NickAragua commented on megamek/src/megamek/common/actions/PushAttackAction.java:
```diff
@@ -46,35 +46,27 @@ public ToHitData toHit(Game game) {
-            return "attacker is evading.";
-        }
-
-        if (!game.getOptions().booleanOption(OptionsConstants.BASE_FRIENDLY_FIRE)) {
```
Removed this because it's being done in the base class already.

NickAragua commented on megamek/src/megamek/common/actions/PushAttackAction.java:
```diff
@@ -46,35 +46,27 @@ public ToHitData toHit(Game game) {
      */
     protected static String toHitIsImpossible(Game game, Entity ae, Targetable target) {
         String physicalImpossible = PhysicalAttackAction.toHitIsImpossible(game, ae, target);
-        String extendedBladeImpossible = null;
```
Removed this because it's being done in the base class already.

repligator commented:
NPE when creating a 1,000 ton object.
[megamek.log](https://github.com/user-attachments/files/16313719/megamek.log)

repligator commented:
A "So-and-so" has fled the map, taking (Object) with it message would be nice.

repligator commented:
If certain long/complicated texts are pasted into the item name field (for example, an entire Wikipedia article) the player settings will go blank.
![Screenshot_20240720_134253](https://github.com/user-attachments/assets/02e15079-fb57-49fd-b6fa-0039fa879485)

The tonnage field will accept negative numbers, although this seems to have no affect on gameplay.

Everything else looks good.

NickAragua commented:
Heh, guess I better put a character limit on the names.

HammerGS commented:
Related: https://github.com/MegaMek/megamek/issues/2577

NickAragua commented:
Updated with testing results, will wait for approval/any other reviews.

repligator commented:
It looks like invincible objects are still getting destroyed under certain circumstances. In this case, the carrier was delimbed. It didn't generate any sort of "object destroyed" message either.
[gamelog.zip](https://github.com/user-attachments/files/16338950/gamelog.zip)
[megamek.log](https://github.com/user-attachments/files/16338952/megamek.log)
[where-is-the-ball.sav.gz](https://github.com/user-attachments/files/16338967/where-is-the-ball.sav.gz)

NickAragua commented:
Looks like the Dasher fell in round 4 due to leg loss which resulted in the ball being dropped but not put on the ground. I'll add some messaging and review why it didn't show up on the ground.

NickAragua commented:
Added a report for when cargo is dropped as a result of a fall.

repligator commented:
Another disappearing invincible object. Last seen being carried by `ID: 4` . Not sure if the disappearance was caused by the destruction of the mech, or the fact the pilot ejected.
[gamelog_1.zip](https://github.com/user-attachments/files/16363635/gamelog_1.zip)
[megamek.log](https://github.com/user-attachments/files/16363636/megamek.log)
[Uploading where_is_the_ball_2.sav.gz…]()

NickAragua commented:
Yep, missed that "workflow". Cargo is now dropped whenever an entity is destroyed for any reason (ejection, etc).

Note that cargo that winds up off the board (via push or skid or otherwise) isn't coming back.

SJuliez commented on megamek/src/megamek/common/Game.java:
```diff
@@ -3273,8 +3278,77 @@ public Map<String, BehaviorSettings> getBotSettings() {
+     * Place a carryable object on the ground at the given coordinates
+     */
+    public void placeGroundObject(Coords coords, ICarryable carryable) {
+    	if (!groundObjects.containsKey(coords)) {
```
The entire method code can be written as
`groundObjects.computeIfAbsent(coords, k -> new ArrayList<>()).add(carryable);`
(not tested)

SJuliez commented on megamek/src/megamek/common/ICarryable.java:
```diff
@@ -0,0 +1,30 @@
+/**
+ * An interface defining all the required properties of a carryable object.
+ */
+public interface ICarryable {
```
I had the hope at some point for something like this to be placed under (i.e. extend) BTObject to be in a neat hierarchy (fwiw). It would then be the first place to add this override:
```
default boolean isCarryableObject() {
    return true;
}
```
It could add a few overrides to return false to the other checks else but that's not terribly important.
It would make sense to change getName() to generalName() (I was out of ideas for a better term) which makes it the equivalent of getChassis().

(Also... it could be named Carryable without the I as there's no old Carryable class in the way)

NickAragua replied:
Didn't realize that there was a BTObject. I'll investigate how practical it'll be to update the interface to fit into the hierarchy.

The I in front of the name follows the general convention (and in MM as well to some extent) for interfaces; it's easier to tell at a glance that a piece of code is an interface if it has the I in front of it. Although I guess with the "default" functions nowadays the difference is a bit blurry.

SJuliez commented:
It is possible to create an EntityImage for Carryables and thus allow it to apply a Camo (the player color e.g. or a fallback brown) and even give them a facing (to rotate them). They could also have smoke :) (no, that doesnt make much sense)
It takes a bit of fluffy code to create it and get an image as a result, that would go into the GroundObjectSprite. I can figure that out if this sounds like a good idea.

NickAragua commented:
Refactored the ICarryable interface so it's part of the BTObject hierarchy.

I'm kind of ambivalent re: camo, as the cargo may not necessarily belong to the player (i.e. scenario rules call for "attacker nominates X squares that contain cargo"). The one thing that I do want to ideally do is have the cargo sprite appear "under" the units rather than over (it looks odd). Any advice on that?

SJuliez commented:
Ah ok, I didn't know if they should be player-colored. My intention was more to not keep them this black/white. Giving them a fixed brownish camo (we sure have one) is a simple option. But we can leave that for any point later.

Sprite ordering ... is not in an ideal place. There is Sprite.getSpritePriority() but it is only used with the xyzSpriteHandlers while the older code just draws the sprites according to type. Since the handler sprites are mostly the ones that go over everything (move envelope) they currently draw after the units and it wouldnt be helpful to swap that. I'll take a look if the unit sprites can be added to the allSprites treemap without a big hassle so they would respect draw order, then the sprite priority should help. Will be back on this.

repligator commented:
As of the latest push, the object names are displaying like this
![Screenshot_20240726_085242](https://github.com/user-attachments/assets/632fd865-ef78-446f-8c64-3d742a9ca16a)

NickAragua commented:
Should be better now. Got a little too overzealous in my refactoring.

SJuliez commented:
I have created #5801
That one injects the entity sprites into the AbstractBoardView sprite drawing process and so allows the Carryable sprites to draw under units. You can pull it into this one, if you want. I merged them locally and it seemed to work fine.

NickAragua commented:
Ok, merged #5801 into this one. I think this one's ready to go now.

SJuliez commented:
The briefcase sprite doesnt scale with the board zoom.

Suggestion: replace the empty prepare() in GroundObjectSprite with this:

```
    @Override
    public void prepare() {
        updateBounds();
        image = createNewHexImage();
        Graphics2D graph = (Graphics2D) image.getGraphics();
        UIUtil.setHighQualityRendering(graph);
        graph.scale(bv.scale, bv.scale);
        graph.drawImage(CARGO_IMAGE, 0, 0, null);
    }
```

:)

repligator commented:
The report for a dropped object seems to get nested under the last physical attack. ID:95 was the one carrying the object.
![Screenshot_20240728_064022](https://github.com/user-attachments/assets/ee5d0911-e986-4a01-a679-287327ef53fb)

SJuliez commented:
I'm getting problems when trying to pick up stacked objects, i.e. more than 1 in a hex, specifically a light 2t and a heavy 30t. The game lets me pick up one, no questions asked (as the 30t is too heavy) but it does not really pick it up.

SJuliez commented:
Would it make sense to show carried objects in the unit tooltip? (or does it already?)

SJuliez commented:
Hoping I'm not missing something that's there already, I feel that it's likely we will at some point want to be able to identify specific carryables and maybe have them have an owner (thinking of capture the flag). May I suggest having ICarryable extend InGameObject instead of BTObject and adding the following to Briefcase.
As currently owner and id arent used anywhere, no other changes are necessary but this adds the infrastructure.


```
	private int id;
	private int ownerId;
	
	@Override
	public int getId() {
		return id;
	}

	@Override
	public void setId(int newId) {
		this.id = newId;
	}

	@Override
	public int getOwnerId() {
		return ownerId;
	}

	@Override
	public void setOwnerId(int newOwnerId) {
		this.ownerId = newOwnerId;
	}

	@Override
	public int getStrength() {
		return 0;
	}
```

NickAragua commented:
Added sprite scaling and unit tooltip; adjusted the ordering of "cargo dropped" messages during PSR resolution; adjusted object hierarchy; make sure to pick up the correct object when the "default" object being picked up isn't the first in the list.
